tests/h/services/group_test.py

```diff
@@ -5,344 +5,408 @@
 
-    def test_member_leave_removes_user_from_group(self, service, users):
-        group = Group(name='Theresa and her buddies',
-                      authority='foobar.com',
-                      creator=users['theresa'])
-        group.members.append(users['cazimir'])
+        svc.member_leave(group, new_member.userid)
+        svc.member_leave(group, new_member.userid)
 
-        service.member_leave(group, 'cazimir')
+        assert new_member not in group.members
 
-        assert users['cazimir'] not in group.members
+    def test_member_leave_publishes_leave_event(self, svc, factories, publish):
+        group = factories.Group()
+        new_member = factories.User()
+        group.members.append(new_member)
 
-    def test_member_leave_is_idempotent(self, service, users):
-        group = Group(name='Theresa and her buddies',
-                      authority='foobar.com',
-                      creator=users['theresa'])
-        group.members.append(users['cazimir'])
+        svc.member_leave(group, new_member.userid)
 
-        service.member_leave(group, 'cazimir')
-        service.member_leave(group, 'cazimir')
+        assert publish.called_once_with('group-leave', group.pubid, new_member.userid)
 
-        assert users['cazimir'] not in group.members
 
-    def test_member_leave_publishes_leave_event(self, service, users, publish):
-        group = Group(name='Donkey Trust',
-                      authority='foobari.com',
-                      creator=users['theresa'])
-        group.members.append(users['cazimir'])
-        group.pubid = 'abc123'
+class TestGroupServiceUpdateMembership(object):
 
-        service.member_leave(group, 'cazimir')
+    def test_update_membership_adds_users_in_usernames(self, factories, svc):
+        group = factories.OpenGroup()  # no members at outset
+        new_members = [
+            factories.User(),
+            factories.User()
+        ]
 
-        publish.assert_called_once_with('group-leave', 'abc123', 'cazimir')
+        svc.update_membership(group, [user.username for user in new_members])
 
-    def test_update_memberships_adds_and_remove_members(self, factories, publish, service, users):
-        group = Group(name='Donkey Trust',
-                      authority='example.com',
-                      creator=users['theresa'])
-        group.pubid = 'abc123'
-        service.member_join = mock.Mock()
-        service.member_leave = mock.Mock()
+        # The following is the desired end state but the ordering is currently non-deterministic
+        # TODO refactor method such that ordering is deterministic
+        # assert group.members == new_members
 
-        group.members.append(users['cazimir'])
-        group.members.append(users['theresa'])
-        group.members.append(users['ali'])
+        for member in group.members:
+            assert member in new_members
+        for member in new_members:
+            assert member in group.members
```

Comment:
There's also a thing called `matchers.unordered_list`: `assert group.members == matchers.unordered_list(new_members)`

Reply:
(I'm not sure that it's _only_ the use of `set` in the groups service that means the order of members is non-deterministic. Is the order of `Group.members` at the model level deterministic? I don't think postgres queries are ordered unless you specifically order them by some criteria.)